Guoweikang/linked list

guoweikang opened the pull request:
Update LinkedList to make it easier to use
    
    --------
    1 add List sample code
    2 add more test case
    3 remove iter_back,use iter().rev()

guoweikang committed:
Synchronize r4l linked_lists latest code

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
Update LinkedList to make it easier to use

--------
1 add List sample code
2 add more test case
3 remove iter_back,use iter().rev()

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
remove unused unsafe_list source code

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
add macro def_node and def_generic_node

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
fix clippy warnning

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
fix doc warning

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
add inner() func for node

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
create tag and update version

--------
crate tag v0.2.0

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
add doc for pub func

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
inner not a pub filed, use inner() api access

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

equation314 commented on src/lib.rs:
```diff
@@ -3,12 +3,187 @@
+
+            #[inline]
+            #[doc = "Get inner"]
+            pub fn inner(&self) -> &$type {
```
const fn

guoweikang replied:
In prev version : use crate::unsafe_list::{self, Adapter, Cursor, Links};  function param &self not means the ref of instances , otherwise it is unsafe_list::self type ;

Current, &self not a type, it is RawList instances ref

equation314 commented on src/linked_list.rs:
```diff
@@ -84,62 +83,64 @@ impl<T: ?Sized> Wrapper<T> for &T {
     /// Returns whether the list is empty.
-    #[inline]
-    pub const fn is_empty(&self) -> bool {
+    pub fn is_empty(&self) -> bool {
```
Why not use const fn?

guoweikang replied:
In prev version : `use crate::unsafe_list::{self, Adapter, Cursor, Links}; `   function param  `&self` not means the  ref of instances , otherwise it is unsafe_list::self type ; 

Current, `&self`  not a type, it is RawList instances ref

equation314 commented on src/lib.rs:
```diff
@@ -3,12 +3,187 @@
+
+            #[inline]
+            #[doc = "Get inner"]
+            pub fn inner(&self) -> &T {
```
const fn

equation314 commented on src/lib.rs:
```diff
@@ -3,12 +3,187 @@
+
+#[allow(missing_docs)]
+#[macro_export]
+macro_rules! def_node {
```
Add comments here to generate better rustdoc

guoweikang replied:
fixed

guoweikang committed:
Add DOC and example for def_node macro

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
use const fn for inner and is_empty

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang reviewed (commented):
OK

equation314 commented on src/lib.rs:
```diff
@@ -3,12 +3,205 @@
 //!
 //! [1]: https://github.com/Rust-for-Linux/linux/blob/rust/rust/kernel/linked_list.rs
+//!
+//! In more general use cases, shoud not use RawList directly,
```
RawList -> ``[`RawList`]``, same with List below.

equation314 commented on src/lib.rs:
```diff
@@ -3,12 +3,205 @@
+#[macro_export]
+macro_rules! def_node {
+    ($struct_name:ident, $type:ty) => {
+        #[doc = "A node wrapper for inner type "]
```
How to deal with the user-defined doc of the node?

equation314 replied:
I suggest this api style:
```rs
define_node! {
    /// A node with usize value.
    struct UsizeNode(usize);
    /// A node with inner type.
    struct WrapperNode<T>(T);
}

guoweikang replied:
fix： now we can use like this : 
```rust
def_node!(
     /// An example Node with usize
     struct ExampleNode1(usize);
     /// An example Node with generic Inner type
     struct GenericNode1<T>(T);
);
```
can create : 
```rust
/// An example Node with usize
struct ExampleNode1 {
    inner: usize,
    links: crate::Links<Self>,
}
impl crate::GetLinks for ExampleNode1 {
    type EntryType = Self;
    #[inline]
    fn get_links(t: &Self) -> &crate::Links<Self> {
        &t.links
    }
}
impl ExampleNode1 {
    ///Create a node
    const fn new(inner: usize) -> Self {
        Self {
            inner,
            links: crate::Links::new(),
        }
    }
    #[inline]
    ///Get inner
    const fn inner(&self) -> &usize {
        &self.inner
    }
}
impl core::ops::Deref for ExampleNode1 {
    type Target = usize;
    #[inline]
    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
/// An example Node with generic Inner type
struct GenericNode1<T> {
    inner: T,
    links: crate::Links<Self>,
}
impl<T> crate::GetLinks for GenericNode1<T> {
    type EntryType = Self;
    #[inline]
    fn get_links(t: &Self) -> &crate::Links<Self> {
        &t.links
    }
}
impl<T> GenericNode1<T> {
    ///Create a node
    const fn new(inner: T) -> Self {
        Self {
            inner,
            links: crate::Links::new(),
        }
    }
    #[inline]
    ///Get inner
    const fn inner(&self) -> &T {
        &self.inner
    }
}
impl<T> core::ops::Deref for GenericNode1<T> {
    type Target = T;
    #[inline]
    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
```

equation314 commented on src/linked_list.rs:
```diff
@@ -84,61 +83,63 @@ impl<T: ?Sized> Wrapper<T> for &T {
+impl<T: GetLinks + ?Sized> GetLinks for Box<T> {
     type EntryType = T::EntryType;
-
-    #[inline]
```
Missing inline

equation314 commented on src/linked_list.rs:
```diff
@@ -162,59 +167,172 @@ impl<G: AdapterWrapped> List<G> {
+}
+
+impl<'a, G: GetLinksWrapped> Iterator<'a, G> {
+    fn new(list: &'a List<G>) -> Self {
```
const fn

equation314 commented on src/raw_list.rs:
```diff
@@ -0,0 +1,562 @@
+}
+
+impl<'a, G: GetLinks> Iterator<'a, G> {
+    fn new(cursor_front: Cursor<'a, G>, cursor_back: Cursor<'a, G>) -> Self {
```
const fn

guoweikang replied:
![1729477374132](https://github.com/user-attachments/assets/0042ad7c-4162-4f9d-bcef-9267500bd262)
not all funcs allowed  add const.

equation314 commented on src/raw_list.rs:
```diff
@@ -0,0 +1,562 @@
+}
+
+impl<'a, G: GetLinks> CursorMut<'a, G> {
+    fn new(list: &'a mut RawList<G>, cur: Option<NonNull<G::EntryType>>) -> Self {
```
const fn?

guoweikang replied:
![1729477374132](https://github.com/user-attachments/assets/0042ad7c-4162-4f9d-bcef-9267500bd262)
not all funcs allowed  add const.

equation314 commented on src/linked_list.rs:
```diff
@@ -162,59 +167,172 @@ impl<G: AdapterWrapped> List<G> {
-    pub fn cursor_front(&self) -> Cursor<'_, G> {
-        self.list.cursor_front()
+    /// Returns a mutable cursor starting on the first (front) element of the list.
+    pub fn cursor_front_mut(&mut self) -> CursorMut<'_, G> {
```
Why no cursor_front for read only use?

guoweikang replied:
because now we support iterator， no need to supply cursor

guoweikang committed:
macro def_node with new format

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
remove unsed description

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
support pub(crate) for def_node

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang commented on src/raw_list.rs:
```diff
@@ -0,0 +1,562 @@
+}
+
+impl<'a, G: GetLinks> CursorMut<'a, G> {
+    fn new(list: &'a mut RawList<G>, cur: Option<NonNull<G::EntryType>>) -> Self {
```
error[E0658]: mutable references are not allowed in constant functions

guoweikang commented on src/linked_list.rs:
```diff
@@ -162,59 +169,172 @@ impl<G: AdapterWrapped> List<G> {
 }
+
+/// A list cursor that allows traversing a linked list and inspecting & mutating elements.
+pub struct CursorMut<'a, G: GetLinksWrapped> {
```
because I add iterfator（wrpaaed  a `rawlist Cursor` ） for linkedlist

guoweikang committed:
use const for more func

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
use vis in macro

--------
1 use vis in macro
2 replace def_node node(type) with node{type}

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

equation314 commented on src/lib.rs:
```diff
@@ -1,14 +1,161 @@
+#[macro_export(local_inner_macros)]
+#[doc(hidden)]
+macro_rules! __def_node_internal {
+    ($(#[$meta:meta])* ($($vis:tt)*) struct $name:ident{$type:ty};) => {
```
`$vis:tt` -> `$vis:vis`

guoweikang replied:
done

equation314 commented on src/lib.rs:
```diff
@@ -1,14 +1,161 @@
+            $($vis)* const fn inner(&self) -> &$type {
+                &self.inner
+            }
+        }
```
Add `into_inner`, see https://github.com/arceos-org/arceos/pull/184#discussion_r1806679269

guoweikang replied:
Ok, had add into_inner func for node def

equation314 commented on src/lib.rs:
```diff
@@ -1,14 +1,161 @@
+///
+#[macro_export(local_inner_macros)]
+macro_rules! def_node {
+    ($(#[$meta:meta])* $sv:vis struct $name:ident{$type:ty}; $($t:tt)*) => {
```
`$name:ident{$type:ty}` -> `$name:ident($type:ty)`

guoweikang replied:
done

equation314 commented on src/lib.rs:
```diff
@@ -1,14 +1,161 @@
+///
+#[macro_export(local_inner_macros)]
+macro_rules! def_node {
+    ($(#[$meta:meta])* $sv:vis struct $name:ident{$type:ty}; $($t:tt)*) => {
```
`$sv` -> `$vis`

guoweikang replied:
done

equation314 commented on src/lib.rs:
```diff
@@ -1,14 +1,161 @@
+    ($(#[$meta:meta])* $sv:vis struct $name:ident{$type:ty}; $($t:tt)*) => {
+        __def_node_internal!($(#[$meta])* ($sv) struct $name{$type};);
+        def_node!($($t)*);
+
```
Remove blank line

equation314 commented on README.md:
```diff
@@ -0,0 +1,60 @@
+# LinkedList
+
```
Add CI and doc badge

guoweikang replied:
Fixed  now linked_list are ready to pushed crates-io

equation314 commented on README.md:
```diff
@@ -0,0 +1,60 @@
+ list.push_back(node1);
+ list.push_back(node2);
+
+ //Support Iter
```
Add a space after `//`

guoweikang replied:
fixed

guoweikang committed:
Add CI DOC for linked list

--------
1 linked list push to crates-io
2 add into_inner func for def_node

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
modify crates-io package name from linked_list to linked_list_r4l

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
fix cargo fmt

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
fix unitest fail

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
update package version to 0.2.0

-------
update version for publish to craes-io

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
use vis in  _def_node_inernal

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

equation314 commented on src/lib.rs:
```diff
@@ -1,14 +1,178 @@
+
+        impl $name {
+            #[doc = "Create a node"]
+            $vis const fn new(inner: $type) -> Self {
```
Always use `pub` in the impl block

guoweikang replied:
done

equation314 commented on src/lib.rs:
```diff
@@ -1,14 +1,178 @@
+
+            #[inline]
+            #[doc = "Consumes the `node`, returning the wrapped inner"]
+            $vis fn into_inner(self) -> $type {
```
const fn

guoweikang replied:
so, in generic type it would complain ` error[E0493]: destructor of `GenericNode<T>` cannot be evaluated at compile-time`  

add for normal node

equation314 commented on Cargo.toml:
```diff
@@ -1,12 +1,14 @@
 description = "Linked lists that supports arbitrary removal in constant time"
 license = "GPL-2.0-or-later"
 homepage = "https://github.com/arceos-org/arceos"
 repository = "https://github.com/arceos-org/linked_list"
```
Rename this repo to `linked_list_r4l`

guoweikang replied:
Ok, but renaming the repository requires you to complete this operation

equation314 commented on README.md:
```diff
@@ -0,0 +1,64 @@
+
+[![Crates.io](https://img.shields.io/crates/v/linked_list_r4l)](https://crates.io/crates/linked_list_r4l)
+[![Doc.rs](https://docs.rs/linked_list_r4l/badge.svg)](https://docs.rs/linked_list_r4l)
+[![CI](https://github.com/arceos-org/linked_list/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/arceos-org/linked_list/actions/workflows/ci.yml)
```
Rename this repo to `linked_list_r4l`

guoweikang replied:
Ok, but renaming the repository requires you to complete this operation

guoweikang committed:
rename repo to linked_list_r4l

--------
1 rename repo to linked_list_r4l
2 change to pub impl for node
3 const fn

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>

guoweikang committed:
remove const fn for into_innner

Signed-off-by: guoweikang <guoweikang.kernel@gmail.com>
